Quantities with specific types -> use in Angle, Distance

mhvk opened the pull request:
We have two `Quantity` subclasses that are for specific physical types, `Angle` and `Distance`. These are each set up such that if operations change the unit, they loose their type. Currently, this has some loop-holes, however (which led me to label this as a bug, since it is unintended):
```
In [5]: a = Angle(10., 'deg')

In [6]: a *= 'm'

In [7]: a
Out[7]: <Angle 10.0 deg m>

In [8]: (10. * u.m).view(Angle)
Out[8]: <Angle 10.0 m>
```
The first commit in this PR closes these loopholes by ensuring that all setting of units goes through a single `_set_unit` method, which checks that a given unit is consistent with the class. It also moves common code from `Angle` and `Distance` to a new `SpecificTypeQuantity` subclass

Note that I also use this to also define a `DimensionlessQuantity` class, whose main difference from other classes is that if one does `array.view(DimensionlessQuantity)` the unit will automatically be dimensionless (rather than unset). I can remove this if need be to keep this just a pure bug-fix PR.

The second commit uses the new `_set_unit` also for function quantities (magnitudes, etc.), which turns out to rather greatly simplify the code there. As part of this, it now also ensures that those can only be viewed as regular quantities if their physical dimension is dimensionless unscaled.

cc @eteq (as it touches coordinates) and @astrofrog.

adrn commented on astropy/units/quantity.py:
```diff
@@ -576,40 +587,69 @@ def _new_view(self, obj, unit=None):
+                    .format(type(self).__name__, UnitBase, type(unit)))
+
+        self._unit = unit
+        return self
```
[very minor question] 
I'm curious why you return `self` here -- since this will always be called as `self._set_unit(...)` the instance should be on hand already, right?

mhvk replied:
It is to be able to do `return self._set_unit(...)`, but I was really unsure whether to do this or not. Given that you are confused, let me undo that.

adrn commented on astropy/units/core.py:
```diff
@@ -461,6 +461,11 @@ class UnitConversionError(UnitsError, ValueError):
 
+class UnitTypeError(UnitsError, TypeError):
+    """
+    Used specifically for errors in setting to units not allowed by a class.
```
I actually don't think we need a new error for this (just use `UnitsError` with a descriptive msg), but don't feel very strongly. If you want to keep this, the description could be clearer -- I don't really understand what is there now.

mhvk replied:
The only reason I made this is that currently `angle *= u.m` would raise a `TypeError` (in `Quantity.__array_prepare__`, and I felt this was sufficiently similar to `angle.view(Distance)` that it should have the same type of exception.  I realise now, though, that I didn't actually make the change in `__array_prepare__`; will do that now...

adrn commented:
Testing the code you put in the PR text as a demo I get a `UnitTypeError` now as expected:

```ERROR: UnitTypeError: Angle instances require units equivalent to 'rad', so cannot set it to 'm'. [astropy.units.quantity]```

For this type of error, it might be nicer to spell out the long-form name of the unit, e.g.,

```ERROR: UnitTypeError: Angle instances require units equivalent to 'radian', so cannot set it to 'meter'. [astropy.units.quantity]```

for readability -- is there a way to get the long-form name of a unit from the object? In looking for that functionality, I realized something else: for most units, the default name is an abbreviation:

```python
In [15]: u.m.name, u.radian.name, u.rydberg.name, u.angstrom.name
Out[15]: ('m', 'rad', 'Ry', 'Angstrom')

In [16]: u.m.names
Out[16]: ['m', 'meter']

In [17]: u.radian.names
Out[17]: ['rad', 'radian']

In [18]: u.angstrom.names
Out[18]: ['Angstrom', 'AA', 'angstrom']
```

but for angstrom the default name is its proper name -- is this intentional? Sorry these questions are a bit out of scope for this PR, but wanted to make sure I'm not missing something before creating a separate issue to discuss...

adrn commented:
Other than minor comments / discussion above, this looks good to me

mhvk commented:
I like your suggestion about using the long name for the unit in the exception message. I do think the fact that angstrom is not listed first is a mistake; it would need a change in its definition, to list short and long names explicitly:
```
def_unit((['Angstrom', 'AA'], ['Angstrom', 'angstrom']), 0.1 * nm, namespace=_ns,
         doc="ångström: 10 ** -10 m",
         format={'latex': r'\mathring{A}', 'unicode': 'Å',
                 'vounit': 'Angstrom'})
```
I think this is better done as a separate PR.

mhvk commented:
Darn, I forgot: the long names are only defined for individual units, but if a `Quantity` has a composite unit, it wouldn't work to print `self.unit.long_names[0]`. One could go by physical type, but again those are not always defined.

mhvk commented:
I rebased and implemented your comments. I removed `DimensionlessQuantity` for now, since that really is a new feature, not a bug fix (and I think it needs to be documented why it is useful).

adrn commented:
OK, looks good to me!

mhvk commented:
@adrn - thanks for the review! I added a changelog entry and will merge now.

eteq commented:
realized this only now when it merged... but looks good - thanks @mhvk!

eteq commented:
It turns out this is quite a challenge to backport to 1.2.x, and I think it's not clear it's not a behavior change... so I'm re-classifying it 1.3.0 (and will update the changelog appropriately)

mhvk commented:
Totally fine to just have this in 1.3, especially since 1.2.2 is really meant to be a stop-gap with the leap second.
